components/script/dom/globalscope.rs

```diff
@@ -207,6 +290,312 @@ impl GlobalScope {
+        let message_ports = self.message_ports.borrow();
+        let port = message_ports
+            .get(port_id)
+            .expect("Port whose close was called to exist");
+        port.close();
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#message-port-post-message-steps>
+    // Steps 6 and 7
+    pub fn post_messageport_msg(&self, port_id: MessagePortId, task: PortMessageTask) {
+        let message_ports = self.message_ports.borrow();
+        let port = message_ports
+            .get(&port_id)
+            .expect("Port whose postMessage was called to exist");
+        if let Some(entangled_id) = port.entangled_port_id() {
+            // Step 7
+            let this = Trusted::new(&*self);
+            let _ = self.port_message_queue().queue(
+                task!(post_message: move || {
+                    let global = this.root();
+                    // Note: we do this in a task, as this will ensure the global and constellation
+                    // are aware of any transfer that might still take place in the current task.
+                    global.upcast::<GlobalScope>().route_task_to_port(entangled_id, task);
+                }),
+                self,
+            );
+        }
+    }
+
+    /// Route the task to be handled by the relevant port.
+    pub fn route_task_to_port(&self, port_id: MessagePortId, task: PortMessageTask) {
+        if !self.message_ports.borrow().contains_key(&port_id) {
+            // If we don't know about the port,
+            // send the message to the constellation for routing.
+            let _ = self
+                .script_to_constellation_chan()
+                .send(ScriptMsg::RerouteMessagePort(port_id.clone(), task));
+            return;
+        }
+
+        // If the port is not enabled yet, or if it is awaiting the completion of it's transfer,
+        // the task will be buffered and dispatched upon enablement or completion of the transfer.
+        let task_to_dispatch = self
+            .message_ports
+            .borrow_mut()
+            .get_mut(&port_id)
+            .map_or(None, |port| port.handle_incoming(task));
+
+        if let Some(task_to_dispatch) = task_to_dispatch {
+            // Get a corresponding DOM message-port object.
+            let dom_port = self.get_dom_message_port(&port_id);
+
+            let PortMessageTask { origin, data } = task_to_dispatch;
+
+            // Substep 3-4
+            rooted!(in(*self.get_cx()) let mut message_clone = UndefinedValue());
+            if let Ok(ports) = structuredclone::read(self, data, message_clone.handle_mut()) {
+                // Substep 6
+                // Dispatch the event, using the dom message-port.
+                MessageEvent::dispatch_jsval(
+                    &dom_port.upcast(),
+                    self,
+                    message_clone.handle(),
+                    Some(&origin.ascii_serialization()),
+                    None,
```

Comment:
We should be passing an actual source value here, so that the receiver can use `event.source.postMessage`.